src/api/server.rkt

```diff
@@ -53,75 +82,394 @@
 
-; Handles semaphore and async part of a job
-(struct work (id job sema))
-
-(define (run-job job-info)
-  (match-define (work job-id info sema) job-info)
-  (define path (make-path job-id))
-  (cond ;; Check caches if job as already been completed
-    [(hash-has-key? *completed-jobs* job-id) (semaphore-post sema)]
-    [(and (*demo-output*) (directory-exists? (build-path (*demo-output*) path)))
-     (semaphore-post sema)]
-    [else
-     (wrapper-run-herbie info job-id)
-     (hash-remove! *job-status* job-id)
-     (semaphore-post sema)])
-  (hash-remove! *job-sema* job-id))
+(define (backend-improve-result-hash-table backend repr test)
+  (define pcontext (improve-result-pctxs backend))
+
+  (define preprocessing (improve-result-preprocess backend))
+  (define end-hash-table (end-hash (improve-result-end backend) repr preprocessing pcontext test))
+
+  (hasheq 'preprocessing
+          preprocessing
+          'pctxs
+          pcontext
+          'start
+          (improve-result-start backend)
+          'target
+          (improve-result-target backend)
+          'end
+          end-hash-table
+          'bogosity
+          (improve-result-bogosity backend)))
+
+(define (end-hash end repr preprocessing pcontexts test)
+  (define ctx (test-context test))
+  (define-values (processed test-pctx)
+    (for/lists (l1 l2)
+               ([pctx pcontexts])
+               (define-values (train-pcontext test-pcontext) (partition-pcontext pctx))
+               (values (preprocess-pcontext ctx test-pcontext preprocessing) test-pcontext)))
+  (define-values (end-alts train-errors end-errors end-costs)
+    (for/lists (l1 l2 l3 l4)
+               ([analysis end])
+               (match-define (alt-analysis alt train-errors test-errs) analysis)
+               (values alt train-errors test-errs (alt-cost alt repr))))
+  (define fpcores
+    (for/list ([altn end-alts])
+      (~a (program->fpcore (alt-expr altn) (test-context test)))))
+  (define alts-histories
+    (for/list ([alt end-alts] [ppctx processed] [tpctx test-pctx])
+      (render-history alt ppctx tpctx (test-context test))))
+
+  (define vars (test-vars test))
+  (define end-alt (alt-analysis-alt (car end)))
+  (define splitpoints
+    (for/list ([var vars])
+      (define split-var? (equal? var (regime-var end-alt)))
+      (if split-var?
+          (for/list ([val (regime-splitpoints end-alt)])
+            (real->ordinal (repr->real val repr) repr))
+          '())))
+
+  (hasheq 'end-alts
+          fpcores
+          'end-histories
+          alts-histories
+          'end-train-scores
+          train-errors
+          'end-errors
+          end-errors
+          'end-costs
+          end-costs
+          'splitpoints
+          splitpoints))
+
+(define (ctx-hash-table ctx)
+  (hasheq 'vars (context-vars ctx) 'repr (repr-hash-table (context-repr ctx))))
+
+(define (repr-hash-table repr)
+  (hasheq 'name (representation-name repr) 'type (representation-type repr)))
```

Comment:
This is fine but I'd prefer calling these `context->json` and `repr->json`

Reply:
Took me a minute to realize you wanted me to change the names. I was looking for context and repr `->json`.